src/lyrics/lyrics.cc

```diff
@@ -0,0 +1,330 @@
+            update_lyrics_window (g_state.title, g_state.artist, g_state.lyrics);
+            return;
+        }
+    }
+
+    if (! aud_get_bool (CFG_SECTION, "enable-file-provider") || ! file_provider.match (g_state))
+    {
+        if (! g_state.artist || ! g_state.title)
+        {
+            update_lyrics_window_error (_("Missing title and/or artist."));
+            return;
+        }
+
+        auto rsrc = remote_source ();
+        if (rsrc)
+        {
+            rsrc->match (g_state);
+            return;
+        }
+    }
+
+    if (! g_state.lyrics)
+        update_lyrics_window_notfound (g_state);
+}
+
+static void destroy_cb ()
+{
+    g_state.filename = String ();
+    g_state.title = String ();
+    g_state.artist = String ();
+    g_state.lyrics = String ();
+
+    hook_dissociate ("tuple change", (HookFunction) lyrics_playback_began);
+    hook_dissociate ("playback ready", (HookFunction) lyrics_playback_began);
+
+    textview = nullptr;
+    textbuffer = nullptr;
+}
+
+static void append_item_to_menu (GtkWidget * menu, const char * label, GCallback callback, void * data)
+{
+    GtkWidget * menu_item = (label != nullptr)
+     ? gtk_menu_item_new_with_label (label)
+     : gtk_separator_menu_item_new ();
+
+    if (callback)
+        g_signal_connect (menu_item, "activate", callback, data);
+
+    gtk_menu_shell_append ((GtkMenuShell *) menu, menu_item);
+    gtk_widget_show (menu_item);
+}
+
+static void append_separator_to_menu (GtkWidget * menu)
+{
+    append_item_to_menu (menu, nullptr, nullptr, nullptr);
+}
+
+static void edit_lyrics_cb (GtkMenuItem * menu_item, void * data)
+{
+    const char * edit_uri = aud::from_ptr<const char *> (data);
+#ifdef USE_GTK3
+    gtk_show_uri_on_window (nullptr, edit_uri, GDK_CURRENT_TIME, nullptr);
+#else
+    gtk_show_uri (nullptr, edit_uri, GDK_CURRENT_TIME, nullptr);
+#endif
+}
+
+static void save_locally_cb (GtkMenuItem * menu_item, void * data)
+{
+    file_provider.save (g_state);
+}
+
+static void refresh_cb (GtkMenuItem * menu_item, void * data)
+{
+    LyricProvider * remote_provider = (LyricProvider *) data;
+    if (remote_provider)
+        remote_provider->match (g_state);
+}
+
+static void populate_popup_cb (GtkTextView * text_view, GtkWidget * menu, void * data)
+{
+    if (! g_state.artist || ! g_state.title || ! GTK_IS_MENU (menu))
+        return;
+
+    LyricProvider * remote_provider = remote_source ();
+
+    append_separator_to_menu (menu);
+
+    if (g_state.lyrics && g_state.source != LyricsState::Source::Local && ! g_state.error)
+    {
+        if (remote_provider)
+        {
+            String edit_uri = remote_provider->edit_uri (g_state);
+
+            // TODO: Is passing 'edit_uri' to 'edit_lyrics_cb' like this safe?
+            if (edit_uri && edit_uri[0])
+                append_item_to_menu (menu, _("Edit Lyrics ..."),
+                 (GCallback) edit_lyrics_cb, aud::to_ptr ((const char *) edit_uri));
+        }
```

Comment:
How would you pass `edit_uri` to the callback function?

Reply:
I would make a copy of the string with `g_strdup()`, then instead of `g_signal_connect()` use `g_signal_connect_data()` and pass `g_free()` as the `destroy_data` parameter.

https://docs.gtk.org/gobject/func.signal_connect_data.html